### PR TITLE
Building a Reinforcement Learning Environment using OpenAI Gym

### LINKED ISSUE #5894: [Machine Learning] Building a Reinforcement Learning Environment using OpenAI Gym
#### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. 

Building a Reinforcement Learning (RL) Environment using OpenAI Gym

#### Proposed article introduction
*We're looking for the first 2-3 paragraphs of your article that appropriately summarize what your article will be about.*
We will use this writing snippet to help us assess the overall quality before approval so ensure it has been proofread, spellchecked and grammarchecked.

Gym is a toolkit for developing and comparing reinforcement learning algorithms. It supports teaching agents everything from walking to playing games like Pong or Pinball. In this tutorial, we will give you a high-level overview of the toolkit and show you how you can build a custom RL environment which you can use in your project.

#### Key takeaways 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
Use this area to get your ideas down on the bulk of your article or tutorial.

By the end of this tutorial, a reader should be able to understand:
- An overview of OpenAI's Gym
- Implement an example to demonstrate how we can build a custom environment
- Test an RL agent on the built custom environment

#### Article quality
How is your article unique? Tell us what makes your approach different from similar articles that have published on the same topic?
Is yours more in-depth? Does it cover additional topics? Do you provide handy tips or anecdotal advice?

Learning about a Reinforcement Learning toolkit and using the standard OpenAI environments is not enough. A reader ought to understand how they can build their own custom environments. We'll use Python to help us accomplish this task. 

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

- [OpenAI Gym](https://gym.openai.com/)

### PR DESCRIPTION
**Pull request by splashtonia**
This closes #5894

## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by WanjaMIKE** [comment]
Some of your code (apart from regular imports) including variable and function names seem to be showing plagiarism.

[New Scan 6_22 AM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7852338/New.Scan.6_22.AM.Copyleaks.Report.pdf)
.

File: content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+There are a couple of things we need to note before we begin:
+- We want our optimal temperature to be between 37 and 39 degrees Celcius. 
+- The shower length will be 60 seconds. If you've worked with other standard RL environments, you know that they have an episode length. In this case, our episode length will be 60 seconds. This means that the model will try to get into that optimal temperature range within 60 seconds.
+- Our model will be performing three actions; Turn up, leave, and turn down. We can turn our temperature up, turn it down, or leave it as it is. 
+
+Let's begin by installing our dependencies.
+
+### Installing and importing required dependencies
+We will be installing four key dependencies:
+- `TensorFlow` allows us to perform training and inferencing of deep learning models.
```
Comment (Bayler) [inline content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md:49]:
```suggestion
`TensorFlow` allows us to perform training and inference on deep learning models.
```

File: content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+- `numpy` to help us with the math.
+- `random` to allow us to test out our random environment.
+
+### Building the custom RL environment with OpenAI Gym
+We begin by creating a `CustomEnv` class. When we pass `Env` to the `CustomEnv` class, we inherit the methods and properties from the OpenAI Gym environment class.
+
+```python
+class CustomEnv(Env):
+```
+We've gone ahead and implemented four different functions within the `CustomEnv` class. We are creating the `__init__` function to initialize the actions, observations, and episode length that we can take when creating your environment.
```
Comment (Bayler) [inline content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md:77]:
```suggestion
We've gone ahead and implemented four different functions within the `CustomEnv` class. We create the `__init__` function to initialize the actions, observations, and episode length we can take when creating your environment. 
```

File: content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    def __init__(self):
+        self.action_space = Discrete(3)
+        self.observation_space = Box(low=np.array([0]), high=np.array([100]))
+        self.state = 38 + random.randint(-3,3)
+        self.shower_length = 60
+```
+The `step` function defines what we do after we take action. We've set our action value to `-1`. Ideally, this means that:
+- If we apply action (0), and apply `-1`, we get a `-1` value. This action will lower the temperature by `1`.
+- If we apply action (1), and apply `-1`, we get a `0` value. This action will make the temperature stay where it is.
+- If we apply action (2), and apply `-1`, we get a `1` value. This action will increase the temperature by `1`.
```
Comment (Bayler) [inline content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md:93]:
```suggestion
- If we apply action (0) together with `-1`, we get a `-1` value. This action will lower the temperature by `1`.
- If we apply action (1) together with `-1`, we get a `0` value. This action will make the temperature stay where it is.
- If we apply action (2) together with `-1`, we get a `1` value. This action will increase the temperature by `1`.
```

File: content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        else:
+            done = False
+        
+        # Setting the placeholder for info
+        info = {}
+        
+        # Returning the step information
+        return self.state, reward, done, info
+```
+The `render` function is used to visualize your results. However, we won't be using it for this tutorial. But, this is where you'd write the visualization code.
```
Comment (Bayler) [inline content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md:122]: Please rewrite this sentence in active voice.

**Comment by Bayler** [comment]
@splashtonia This is a nice read! Just some minor revisions to iron out some of the grammatical issues. We can move on to final review after that is done.

**Comment by splashtonia** [comment]
@Bayler I'm done making the changes.

**Comment by WanjaMIKE** [comment]
Merged #5966 into master

**Pull request merged**